example_logobject_test.go

```diff
@@ -0,0 +1,97 @@
+/*
+Copyright 2021 The logr Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package logr_test
+
+import (
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"github.com/go-logr/logr/funcr"
+)
+
+// NewStdoutLogger returns a logr.Logger that prints to stdout.
+func NewStdoutLogger() logr.Logger {
+	return funcr.New(func(prefix, args string) {
+		if prefix != "" {
+			_ = fmt.Sprintf("%s: %s\n", prefix, args)
+		} else {
+			fmt.Println(args)
+		}
+	}, funcr.Options{})
+}
+
+// ObjectRef references a Kubernetes object
+type ObjectRef struct {
+	Name      string `json:"name"`
+	Namespace string `json:"namespace,omitempty"`
+	Status    string `json:"status,omitempty"`
+}
+
+func (ref ObjectRef) String() string {
+	var result string
+	if ref.Namespace != "" {
+		result = ref.Namespace + "/"
+	}
+	result += ref.Name
+	if ref.Status != "" {
+		result += "(" + ref.Status + ")"
+	}
+	return result
+}
+
+func (ref ObjectRef) GetLogObject() interface{} {
+	// We implement fmt.Stringer for non-structured logging, but we want the
+	// raw struct when using structured logs.  Some logr implementations call
+	// String if it is present, so we want to convert this struct to something
+	// that doesn't have that method.
+	type forLog ObjectRef // methods do not survive type definitions
+	return forLog(ref)
+}
+
+var _ logr.LogObject = ObjectRef{}
+
+// SimpleObjectRef inherits fields and methods from ObjectRef, but
+// has its own log object implementation which returns a struct with
+// a subset of the fields. The name of the fields are also different
+// in the output (upper case instead of lowercase from the JSON
+// annotation).
+type SimpleObjectRef struct {
+	ObjectRef
+}
+
+func (ref SimpleObjectRef) GetLogObject() interface{} {
+	return struct {
+		Name, Namespace string
+	}{
+		Name:      ref.Name,
+		Namespace: ref.Namespace,
+	}
+}
+
+var _ logr.LogObject = SimpleObjectRef{}
+
+func ExampleLogObject() {
+	pod := ObjectRef{Namespace: "kube-system", Name: "some-pod", Status: "running"}
+	l := NewStdoutLogger()
+	l.Info("as string", "pod", pod.String())
+	l.Info("as struct", "pod", pod)
+	l.Info("simplified", "pod", SimpleObjectRef{ObjectRef: pod})
```

Comment:
As an example, this suggest that SimpleObjectRef would be part of a type's public API, which I don't think is what you meant.

Maybe:

```
l := NewStdoutLogger()

pod1 := ObjectRef{Namespace: "kube-system", Name: "some-pod", Status: "running"}
l.Info("as string", "pod", pod1.String())
l.Info("as struct", "pod", pod1)

pod2 := SimpleObjectRef{Namespace: "kube-system", Name: "some-pod", Status: "running"}
l.Info("as string", "pod", pod2.String())
l.Info("as struct", "pod", pod2)
```

?

That isn't much better - the point here is that these are two different ways to
implement the same API, without the caller having to know.

Can we make 2 whole examples?  I don't know how godoc figures that out...

Reply:
I'm worried that the boilerplate code (type definitions, create logger helper) will either be missing or get repeated if we split this example.

I've changed it so that the two example types are now unrelated. That was just unnecessarily complex.

Reply:
Why speculate when it can be tested in advance with `godoc`? :grin: 

With two example functions in the same file I got what I was worried about: godoc just shows each function body and its output, but none of the type definitions. This isn't very useful.

What works is two files with one example function in each. They can still share the `NewStdoutLogger`, which will be shown for the example file which contains it. That looks like a worthwhile improvement, so I've updated the PR to use that approach.